mantidproject/mantid
# Allowing workspace tabs of Inelastic Elwin interface to add multiple workspaces at once

### Description of work
This PR adds the option to add multiple workspaces at once into the workspace table of the Elwin inelastic interface as requested in #33319.

#### Summary of work

Currently, the ELWIN tab of the Data Manipulation interface is working a bit different when files or workspaces are added. Actually the interface is a tab group with a ``Files`` tab and a ``Workspaces`` tab. It is possible to add multiple files at once from the ``Files`` tab but in order to do the same in the workspaces tab, the ``Add Workspaces`` dialog window has to be invoked multiple times, which makes for a slower workflow. 
In order to add the functionality to add multiple workspaces at once, I have considered that the best option in this case would be to design a new type of ``AddWorkspace`` dialog window that could handle adding multiple workspaces at once, with the possibility to select the indexes of the workspaces that are to be added. In order to do that, I have created a new custom widget that inherits from the `QTableWidget` class and shares some of the functionality of Mantid `WorkspaceSelector` widget (communication with the ADS through POCO notifications. ), which is itself a custom QComboBox.
Although the new dialog window is functionally integrated with the Elwin interface. There's no need to have separate file/workspaces tabs (#36460) and also the added workspaces don't need to be added as multiple rows, but rather one row per workspaces  (#36372 ). This PR is a first step in order to address these two issues.


Fixes #33319. 





### To test:
- Testing the new widget: 
   1. Open 3 workspaces in the ADS with suffix `_red.nxs` or `_sqw.nxs`. (For example: ``iris26173_graphite002_red.nxs``, ``iris26176_graphite002_red.nxs`` and ``iris26173_graphite002_sqw.nxs``). 
   2. Go to `Inelastic->Data Manipulation->Elwin tab` interface. Click on ``Workspace`` tab and on the ``Add Workspace`` button. The multiple workspaces QDialog window should prompt, and the table widget should be filled with the added workspaces.
   3. From the ADS, remove one of the added workspaces, rename other one with a non-allowed suffix, and finally clear the ADS, all these changes should be reflected on table in the Workspace QDialog.
   4. Add files from the ``Browse`` button in the QDialog, see that those files are loaded into the ADS and the table (if suffices are allowed). 
   5. Test buttons on the dialog (``Select All``, ``Unify Range``, ``Reset Range``). 
   6. Select 1 or 2 workspaces from the table an click ``Add Data``. See that the data is added, with the appropriate selected indexes.
- Testing Elwin tab:
   1. Make ``Elwin Tab`` test in https://developer.mantidproject.org/Testing/Inelastic/DataManipulationTests.html








---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
